### PR TITLE
Experimental music restrictions

### PR DESCRIPTION
**Pull request by Eirika**
Add some check about the music which will be played by calling the api

### CODE REVIEW

**Comment by Eirika** [comment]
Before merge, have you some question, proposition, rectification ?

File: music/models.py [hunk 1]
```diff
@@ -16,6 +16,14 @@ def search(self, query):
...
         return Provider.search(query)
 
+    def check_validity(self, id):
+        if self.name == "Youtube":
+            for cls in Source.__subclasses__():
+                if cls.__name__ == self.name:
+                    return cls.check_validity(id)
+
+        return True
+
```
Comment (Eirika) [inline music/models.py]: Not sure about this, need verification !

File: player/models.py [hunk 2]
```diff
@@ -45,29 +45,34 @@ def play(self, music=None):
             events[self.name].cancel()
 
         if music:
+            musicSource = Source.objects.get(name=music.source)
```
Comment (Amoki) [inline player/models.py]: music.source is already a music

File: source_modules/youtube/models.py [hunk 3]
```diff
@@ -28,18 +28,30 @@ def get_info(ids):
...
-            'description': detail["snippet"]["description"][:200] + "...",
-            'thumbnail': detail["snippet"]["thumbnails"]["default"]["url"],
-            'views': detail["statistics"]["viewCount"],
-            'duration': get_time_in_seconds(detail["contentDetails"]["duration"]),
-        }
-        videos.append(detailedVideo)
+        added = True
+        if "regionRestriction" not in detail["contentDetails"]:
+            if "blocked" in detail["contentDetails"]["regionRestriction"]:
+                if 'FR' in detail["contentDetails"]["regionRestriction"]["blocked"]:
```
Comment (Amoki) [inline source_modules/youtube/models.py]: language as a variable in settings (fr by default)

File: source_modules/youtube/models.py [hunk 4]
```diff
@@ -28,18 +28,30 @@ def get_info(ids):
...
+                    added = False
+            if "allowed" in detail["contentDetails"]["regionRestriction"]:
+                if len(detail["contentDetails"]["regionRestriction"]["allowed"]) == 0:
+                    added = False
+                else:
+                    if 'FR' not in detail["contentDetails"]["regionRestriction"]["allowed"]:
```
Comment (Amoki) [inline source_modules/youtube/models.py]: idem

File: source_modules/youtube/models.py [hunk 5]
```diff
@@ -90,3 +102,40 @@ def search(query):
         TemporaryMusic.objects.bulk_create(videos)
 
         return videos
+
+    @staticmethod
+    def check_validity(id):
```
Comment (Amoki) [inline source_modules/youtube/models.py]: Same method for soundcloud that return always true.
Like a JAVA interface

File: music/models.py [hunk 6]
```diff
@@ -16,6 +16,14 @@ def search(self, query):
 
         return Provider.search(query)
 
+    def check_validity(self, id):
+        if self.name == "Youtube":
```
Comment (Amoki) [inline music/models.py]: juste remove this line, (see last comment about interfaces)

File: player/models.py [hunk 7]
```diff
@@ -45,29 +45,34 @@ def play(self, music=None):
...
-
-            message = {
-                'action': 'play',
-                'update': True,
-                'source': music.source.name,
-                'options': {
-                    'name': music.name,
-                    'musicId': music.music_id,
-                    'timer_start': music.timer_start,
+            if musicSource.check_validity(music.music_id):
```
Comment (Amoki) [inline player/models.py]: check this on the end point

File: source_modules/youtube/models.py [hunk 8]
```diff
@@ -90,3 +102,40 @@ def search(query):
...
+            country_validity = True
+            if "regionRestriction" in detail["items"][0]["contentDetails"]:
+                if "blocked" in detail["items"][0]["contentDetails"]["regionRestriction"]:
+                    if 'FR' in detail["items"][0]["contentDetails"]["regionRestriction"]["blocked"]:
+                        country_validity = False
+                if "allowed" in detail["items"][0]["contentDetails"]["regionRestriction"]:
+                    if len(detail["items"][0]["contentDetails"]["regionRestriction"]["allowed"]) == 0:
+                        country_validity = False
+                    else:
+                        if 'FR' not in detail["contentDetails"]["regionRestriction"]["allowed"]:
```
Comment (Amoki) [inline source_modules/youtube/models.py]: idem

File: source_modules/youtube/models.py [hunk 9]
```diff
@@ -90,3 +102,40 @@ def search(query):
...
+        # General validity
+        validity = True
+
+        # Check if the video exist
+        if detail["pageInfo"]["totalResults"] > 0:
+            # Check if the music have a Country restriction
+            country_validity = True
+            if "regionRestriction" in detail["items"][0]["contentDetails"]:
+                if "blocked" in detail["items"][0]["contentDetails"]["regionRestriction"]:
+                    if 'FR' in detail["items"][0]["contentDetails"]["regionRestriction"]["blocked"]:
```
Comment (Amoki) [inline source_modules/youtube/models.py]: idem

File: player/models.py [hunk 10]
```diff
@@ -47,27 +47,30 @@ def play(self, music=None):
...
-
-            message = {
-                'action': 'play',
-                'update': True,
-                'source': music.source.name,
-                'options': {
-                    'name': music.name,
-                    'musicId': music.music_id,
-                    'timer_start': music.timer_start,
+            if music.source.check_validity(music.music_id):
```
Comment (Amoki) [inline player/models.py]: check in the endpoint

Comment (Amoki) [inline player/models.py]: add a convinience shortcut with a music.checkValidity() function

**Pull request merged**